components/org.wso2.carbon.aws/src/main/java/org/wso2/carbon/aws/user/store/mgt/util/AWSSignatureV4Generator.java

```diff
@@ -0,0 +1,424 @@
+    }
+
+    /**
+     * Generate AWS signature key.
+     *
+     * @param key         Key to be used for signing.
+     * @param date        Current date stamp.
+     * @param regionName  Region name of AWS cloud directory.
+     * @param serviceName Name of the service being addressed.
+     * @return Signature key.
+     * @throws UnsupportedEncodingException The Character Encoding is not supported.
+     * @throws InvalidKeyException          This is the exception for invalid Keys (invalid encoding, wrong length,
+     *                                      uninitialized, etc).
+     * @throws NoSuchAlgorithmException     When a particular cryptographic algorithm that is requested is not available.
+     */
+    private byte[] getSignatureKey(String key, String date, String regionName, String serviceName)
+            throws UnsupportedEncodingException, InvalidKeyException, NoSuchAlgorithmException {
+
+        byte[] kSecret = (AWSConstants.AWS4 + key).getBytes(AWSConstants.UTF_8);
+        byte[] kDate = hmacSHA256(kSecret, date);
+        byte[] kRegion = hmacSHA256(kDate, regionName);
+        byte[] kService = hmacSHA256(kRegion, serviceName);
+        return hmacSHA256(kService, AWSConstants.AWS4_REQUEST);
+    }
+
+    /**
+     * Convert byte array to Hex.
+     *
+     * @param bytes bytes to be hex encoded.
+     * @return hex encoded String of the given byte array.
+     */
+    private String bytesToHex(byte[] bytes) {
+
+        final char[] hexArray = AWSConstants.HEX_ARRAY_STRING.toCharArray();
+        char[] hexChars = new char[bytes.length * 2];
+        for (int j = 0; j < bytes.length; j++) {
+            int v = bytes[j] & 0xFF;
+            hexChars[j * 2] = hexArray[v >>> 4];
+            hexChars[j * 2 + 1] = hexArray[v & 0x0F];
+        }
+        return new String(hexChars).toLowerCase();
+    }
+
+    /**
+     * Get timestamp with yyyyMMdd'T'HHmmss'Z' format.
+     *
+     * @return Time stamp value.
+     */
+    private String getTimeStamp() {
+
+        DateFormat dateFormat = new SimpleDateFormat(AWSConstants.DATE_TIME_FORMAT);
+        dateFormat.setTimeZone(TimeZone.getTimeZone(AWSConstants.UTC));//server timezone
+        return dateFormat.format(new Date());
+    }
+
+    /**
+     * Get date with yyyyMMdd format.
+     *
+     * @return Date.
+     */
+    private String getDate() {
+
+        DateFormat dateFormat = new SimpleDateFormat(AWSConstants.DATE_FORMAT);
+        dateFormat.setTimeZone(TimeZone.getTimeZone(AWSConstants.UTC));//server timezone
+        return dateFormat.format(new Date());
+    }
+
+    /**
+     * Encode string value.
+     *
+     * @param param String value that need to be encode.
+     * @return encoded string.
+     */
+    private String encodeParameter(String param) {
+
+        try {
+            return URLEncoder.encode(param, AWSConstants.UTF_8);
+        } catch (UnsupportedEncodingException e) {
+            return URLEncoder.encode(param);
```

Comment:
Upon failing encoding using AWSConstants.UTF_8 why do we fall back to this encoding ? Better to add a comment reasoning out.